docs/architecture/clients/dependency-injection.md

```diff
@@ -0,0 +1,233 @@
+---
+sidebar_position: 4
+---
+
+# Dependency Injection
+
+Dependency injection (DI) in our Typescript clients is manually configured across several locations.
+
+You will need to update the DI configuration when you create a new service or change the
+dependencies injected into an existing service.
+
+## Angular and Non-Angular contexts
+
+Our Typescript clients use both Angular and non-Angular contexts. This can vary:
+
+- between apps - for example, the web client is entirely Angular, whereas the CLI client does not
+  use Angular at all
+- within apps - for example, the browser client uses Angular in its popup but not in its background
+  page/service worker
+
+This has important consequences for how we write services and how we configure DI. In particular:
+
+- Angular contexts use the Angular dependency injection framework
+- non-Angular contexts do not use any dependency injection framework and must manually instantiate
+  their dependencies
+
+## Services
+
+Services should be organized as follows depending on where they're used:
+
+- `libs/common` for services used by **multiple clients** in **non-Angular contexts (or a mix of
+  Angular and non-Angular contexts)**
+- `libs/angular` for services used by **multiple clients** in **Angular contexts only**
+- `apps/<client-name>` for services used by a single client only
+
+Some teams also have their own `libs` folder, which are structured similarly. e.g. `libs/auth`,
+containing `libs/auth/common` and `libs/auth/angular`. These should be used if available.
+
+If a service is used in an Angular context only, it can use the Angular `@Injectable` decorator to
+automatically configure its dependencies. If a service is used in mixed contexts (e.g.
+`libs/common`), it must not use Angular decorators and its dependencies must be manually configured.
+
+## Interfaces
+
+A service should implement an interface defined by an abstract class if:
+
+1. The service will be used across multiple client apps, or
+2. You want to define multiple interfaces for your service to limit what functionality is available
+   to different consumers
```

Comment:
The purpose of this para isn't to explain TS interfaces vs. abstract classes, which is why I didn't go into that detail. It's solely about your second paragraph.

Jest doesn't require interfaces for mocks, so they're not strictly required for unit testing. I understand the more general point about why interfaces are good for loose coupling, but we don't enforce any such requirement today. A regex all-text search in vscode for `export class [A-Z,a-z]*service \{$` returns 61 services that don't `implement` any abstraction. We can change that but it would be a change.

See also the Slack discussion [here](https://bitwarden.slack.com/archives/C02LPDKPML6/p1712797671412329).

This para is basically trying to come up with some clear rules for why we use abstract classes sometimes but not all the time.

Reply:
good points and discussion, thanks! I agree with your summary, now